### DISCUSSION TITLE
Render layer to offscreen framebuffer and use it as a texture for another layer

### DISCUSSION THREAD

**Discussion by NataliaDSmirnova** [visgl/deck.gl#6239] [Q&A] [answered]
Hello guys,

Your library is great for rapid development of layered data visualization, thanks for it  🧡

I get stuck with unusual problem. I need to render CompositeLayer with all its children to offscreen buffer and then use the buffer as a texture for another layer. The action should be done every frame.

I expected to find something like Layer.onBeforeRender/Layer.onAfterRender as a place to insert framebuffer bind/unbind. But I didn't find it. 
- Deck.onBeforeRender exists and it doesn't match my requirements because I need to render to off-screen buffer a deck layers subset only, not all layers.

Could you advise some possibility that I've missed?

**Comment by NataliaDSmirnova** [comment]
Is the task rare or too complicated? Any updates?

**Comment by Pessimistress** [comment, marked as answer]
https://deck.gl/docs/api-reference/core/deck#_framebuffer

Example [here](https://github.com/visgl/deck.gl/blob/master/test/apps/framebuffer/app.js)

**Reply by NataliaDSmirnova** [reply]
Thanks a lot for the example. As a understand it just adds framebuffer for deck.gl globally. In the case all actual layers are rendered to the buffer after main rendering step.

So there is no way to render layers subset to the buffer and then use a resulting texture for another subset of layers.

I've implemented that in another way.

**Reply by Pessimistress** [reply]
> render layers subset to the buffer and then use a resulting texture for another subset of layers

```js
const myFBO = new Framebuffer(...);

const deck = new Deck({
  ...
  _customRender: (redrawReason) => {
    if (shouldUpdateFramebuffer(redrawReason)) {
      deck._drawLayers('update framebuffer', {
        target: myFBO,
        layerFilter: ({layer}) => {
          // return true if layer should be drawn to frame buffer
        }
      });
    }
    deck._drawLayers('screen', {
      layerFilter: ({layer}) => {
        // return true if layer should be drawn to render buffer
      }
    });
  }
})
```

See https://github.com/visgl/deck.gl/blob/v8.4.0/modules/core/src/lib/deck.js#L352

**Reply by NataliaDSmirnova** [reply]
Thanks a million for you help.
